### PR TITLE
Introduce the `scan` interface

### PR DESCRIPTION
**Pull request by tengyifei**
We define the interface of the `scan` function and add test to verify its behavior. It is written referencing the JAX implementation.

The `scan` function is currently implemented as a for loop to keep each PR size small. In the next PR, we will replace the forward pass with an optimization implementation that lowers to an HLO `While` op.

### CODE REVIEW

File: test/run_tests.sh [hunk 1]
```diff
@@ -194,6 +194,7 @@ function run_xla_op_tests1 {
 function run_xla_op_tests2 {
   run_test "$CDIR/pjrt/test_dtypes.py"
   run_test "$CDIR/test_while_loop.py"
+  run_test "$CDIR/test_scan.py"
```
Comment (JackCaoG) [inline test/run_tests.sh:197]: maybe add it to the `test/tpu/run_test.sh` as well

File: test/test_scan.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  for i in range(xs_len):
+    carry, y = fn(carry, tree_map(lambda x: x[i], xs))
+    ys.append(y)
+  ys = tree_map(lambda *x: torch.stack(x), *ys)
+  return carry, ys
+
+
+class ScanTest(XlaTestCase):
+
+  def __init__(self, methodName: str = "runTest") -> None:
```
Comment (JackCaoG) [inline test/test_scan.py:26]: what's this `methodName ` used for?

File: torch_xla/experimental/scan.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  if xs_length is None:
+    raise ValueError(f"`xs` {xs} is an empty PyTree.")
+
+  carry = init
+  ys = []
+
+  for i in range(xs_length):
+    carry, y = fn(carry, tree_map(lambda x: x[i], xs))
+    ys.append(y)
```
Comment (JackCaoG) [inline torch_xla/experimental/scan.py:90]: why do we need per layer activation `ys`?

Comment (JackCaoG) [reply torch_xla/experimental/scan.py:90]: is it for the backward? for the user the output of the scan forward should be the last activation, not the activations stacking together,

Comment (tengyifei) [reply torch_xla/experimental/scan.py:90]: > why do we need per layer activation ys?

In the prototype `apply_layers` (public URL: https://colab.research.google.com/drive/18pPBRDPwSE6S9ZzvMkpVJuAlfJ1ezgsl#scrollTo=B6Q49K72Z9V2&line=6&uniqifier=1), I'll be using the `carry` for tracking per layer activation. The `ys` is unused. `apply_layers` currently passes a bunch of zeros and then discards `ys`, thus cutting that part out of the graph. I believe the JAX version of `scan` supports a `fn` returning `None` for this reason. I can look into aligning with that behavior of JAX either as a follow up PR or in this PR depending on your preference.

> is it for the backward?

Based on the previous explanation, I'm currently not using this for backwards. Although now that you mention this, it sounds like an interesting possibility. My design doc contains a description of how backwards is supported in the prototype.

File: test/test_scan.py (hunk 2 above)
Comment (tengyifei) [reply test/test_scan.py:26]: It looks like part of the method signature of a Python `unittest` test case. Although after consulting the docs, looks like we can use `setUp` instead.

**Review by JackCaoG — APPROVED** [review]

**Pull request merged**